### PR TITLE
[Feature] Add segformer decode head and related train config

### PR DESCRIPTION
**Pull request by sennnnn**
- [x]  Add head SegFormerHead;

- [x]  Add some config for segformer;

### CODE REVIEW

**Comment by AncientRemember** [comment]
are the imgsize parameter of MixVisionTransformer is necessary？ segformer drop the position encoding and make the model inductive when test resolution is diff with train

**Comment by Junjun2016** [comment]
> are the imgsize parameter of MixVisionTransformer is necessary？ segformer drop the position encoding and make the model inductive when test resolution is diff with train

In my opinion, the image size is not necessary, we will refactor it.

File: mmseg/models/decode_heads/segformer_head.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        interpolate_mode: The interpolate mode of MLP head upsample operation.
+            Default: 'bilinear'.
+    """
+
+    def __init__(self, interpolate_mode='bilinear', **kwargs):
+        super(SegFormerHead, self).__init__(
+            input_transform='multiple_select', **kwargs)
+
+        self.interpolate_mode = interpolate_mode
+
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/segformer_head.py:24]: Remove some redundant blank lines.

File: mmseg/models/decode_heads/segformer_head.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                    stride=1,
+                    norm_cfg=self.norm_cfg,
+                    act_cfg=self.act_cfg))
+
+        self.conv_fuse = ConvModule(
+            in_channels=self.channels * num_inputs,
+            out_channels=self.channels,
+            kernel_size=1,
+            norm_cfg=self.norm_cfg)
+
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/segformer_head.py:43]: Another suggestion is whether the kernel size of these two ConvModule can be passed in by Args.

**Comment by sennnnn** [comment]
## About the influence of AlignedResize for performance

| Model | Crop Size | Resize Mode | mIoU |
| :-- | :-- | :-- | :-- |
| Segformer_MIT-B0 | 512x512 | Normal | 37.41 | 
| Segformer_MIT-B1 | 512x512 | Normal | 41.05 | 
| Segformer_MIT-B2 | 512x512 | Normal | 45.68 | 
| Segformer_MIT-B0 | 512x512 | Aligned | 38.1 | 
| Segformer_MIT-B1 | 512x512 | Aligned | 41.75 | 
| Segformer_MIT-B2 | 512x512 | Aligned | - |

File: mmseg/models/decode_heads/segformer_head.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from mmcv.cnn import ConvModule
+
+from mmseg.models.builder import HEADS
+from mmseg.models.decode_heads.decode_head import BaseDecodeHead
+from mmseg.ops import resize
+
+
+@HEADS.register_module()
+class SegformerHead(BaseDecodeHead):
+    """The all mlp Head of segformer.
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/segformer_head.py:12]: Add reference. 

Follow the style of other heads.

File: tools/deploy_test.py [hunk 4]
```diff
@@ -79,7 +80,7 @@ def simple_test(self, img: torch.Tensor, img_meta: Iterable,
         if not (ori_shape[0] == seg_pred.shape[-2]
                 and ori_shape[1] == seg_pred.shape[-1]):
             seg_pred = torch.from_numpy(seg_pred).float()
-            seg_pred = torch.nn.functional.interpolate(
+            seg_pred = resize(
```
Comment (xvjiarui) [inline tools/deploy_test.py:83]: Why change this?

File: mmseg/models/decode_heads/segformer_head.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+
+from mmseg.models.builder import HEADS
+from mmseg.models.decode_heads.decode_head import BaseDecodeHead
+from mmseg.ops import resize
+
+
+@HEADS.register_module()
+class SegformerHead(BaseDecodeHead):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/segformer_head.py:11]: Unittest is missing.

File: mmseg/models/decode_heads/segformer_head.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    Args:
+        interpolate_mode: The interpolate mode of MLP head upsample operation.
+            Default: 'bilinear'.
+    """
+
+    def __init__(self, interpolate_mode='bilinear', **kwargs):
+        super().__init__(input_transform='multiple_select', **kwargs)
+
+        self.interpolate_mode = interpolate_mode
+
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/segformer_head.py:23]:
```suggestion
```

File: mmseg/models/decode_heads/segformer_head.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            in_channels=self.channels * num_inputs,
+            out_channels=self.channels,
+            kernel_size=1,
+            norm_cfg=self.norm_cfg)
+
+    def forward(self, inputs):
+        # Receive 4 stage backbone feature map: 1/4, 1/8, 1/16, 1/32
+        inputs = self._transform_inputs(inputs)
+        outs = []
+        for x, conv in zip(inputs, self.convs):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/segformer_head.py:47]: `zip` is not safe since it will stop at the short one. 

Use int to index `x` and `convs`

File: tools/deploy_test.py (hunk 4 above)
Comment (Junjun2016) [reply tools/deploy_test.py]: #731

File: configs/segformer/segformer_mit-b0_512x512_160k_ade20k.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+_base_ = [
+    '../_base_/models/segformer_mit-b0.py', '../_base_/datasets/ade20k.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_160k.py'
+]
+
+model = dict(
+    pretrained='pretrain/mit_b0.pth', decode_head=dict(num_classes=150))
```
Comment (xvjiarui) [inline configs/segformer/segformer_mit-b0_512x512_160k_ade20k.py:7]: update pretrain path

File: mmseg/datasets/pipelines/transforms.py [hunk 9]
```diff
@@ -6,6 +6,74 @@
...
+
+    Args:
+        size_divisor (int): images and gt seg maps need to resize to multiple
+            of size_divisor. Default: 32.
+    """
+
+    def __init__(self, size_divisor=32):
+        self.size_divisor = size_divisor
+
+    def _align(self, img, interpolation=None):
```
Comment (xvjiarui) [inline mmseg/datasets/pipelines/transforms.py:21]: Use https://github.com/open-mmlab/mmcv/blob/571e3e5fc75c23b45cbd9b00011af094357c5f1d/mmcv/image/geometric.py#L98 instead.

File: mmseg/datasets/pipelines/transforms.py [hunk 10]
```diff
@@ -6,6 +6,74 @@
...
+        results['img_shape'] = img.shape
+        results['pad_shape'] = img.shape
+        return img
+
+    def _align_seg(self, results):
+        """Align segmentation map to multiple of size divisor."""
+        for key in results.get('seg_fields', []):
+            gt_seg = results[key]
+            gt_seg = self._align(gt_seg, 'nearest')
+            results[key] = gt_seg
```
Comment (xvjiarui) [inline mmseg/datasets/pipelines/transforms.py:55]: Merge in to call instead.

File: configs/segformer/readme.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+| ------ | -------- | --------- | ------: | ---: | ------------- |
+|Segformer | MIT-B0 | 512x512 | 160000 | 38.1  | 38.57 |
+|Segformer | MIT-B1 | 512x512 | 160000 | 41.64 | 42.76 |
+|Segformer | MIT-B2 | 512x512 | 160000 | 46.53 | 47.49 |
+|Segformer | MIT-B3 | 512x512 | 160000 | 48.46 | 49.14 |
+|Segformer | MIT-B4 | 512x512 | 160000 | 49.34 | 50.29 |
+|Segformer | MIT-B5 | 512x512 | 160000 | 50.08 | 50.72 |
+|Segformer | MIT-B5 | 640x640 | 160000 | - | - |
+
+We replace `AlignedResize` to `Resize + ResizeToMultiple`. If you want to test by
```
Comment (xvjiarui) [inline configs/segformer/readme.md:42]:
```suggestion
We replace `AlignedResize` in original implementatiuon to `Resize + ResizeToMultiple`. If you want to test by
```

File: configs/segformer/readme.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+test_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(
+        type='MultiScaleFlipAug',
+        img_scale=(2048, 512),
+        # img_ratios=[0.5, 0.75, 1.0, 1.25, 1.5, 1.75],
+        flip=False,
+        transforms=[
+            dict(type='Resize', keep_ratio=True),
+            dict(type='ResizeToMultiple', size_divisor=32),
```
Comment (xvjiarui) [inline configs/segformer/readme.md:55]:
```suggestion
            # resize image to multiple of 32, improve SegFormer by xx mIoU. 
            dict(type='ResizeToMultiple', size_divisor=32),
```

Comment (xvjiarui) [reply configs/segformer/readme.md:56]: We may replace xx with a concrete number.

File: configs/segformer/readme.md [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+            dict(type='RandomFlip'),
+            dict(type='Normalize', **img_norm_cfg),
+            dict(type='ImageToTensor', keys=['img']),
+            dict(type='Collect', keys=['img']),
+        ])
+]
+```
+
+Because the MixVisionTransformer is not supported by mmcls, we just can get backbone weights from official repo. We provide convert script to convert official weights in `tools/model_converters/mit_convert.py`. After converting, you need to modify `pretrained` of model config.
```
Comment (xvjiarui) [inline configs/segformer/readme.md:65]:
```suggestion
We convert the backbone weights from the official repo (https://github.com/NVlabs/SegFormer) with `tools/model_converters/mit_convert.py`.
```
Also add cmd example.

File: tools/model_converters/mit_convert.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+    return parser.parse_args()
+
+
+if __name__ == '__main__':
+    args = parse_args()
+    src_path = args.src
+    dst_path = args.dst
+
+    ckpt = torch.load(src_path)
```
Comment (xvjiarui) [inline tools/model_converters/mit_convert.py:72]:
```suggestion
    ckpt = torch.load(src_path, map_location='cpu')
```

File: tools/model_converters/mit_convert.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    if 'state_dict' in ckpt:
+        ckpt = ckpt['state_dict']
+    elif 'model' in ckpt:
+        ckpt = ckpt['model']
+    else:
+        ckpt = ckpt
```
Comment (xvjiarui) [inline tools/model_converters/mit_convert.py:78]: The official weight should only contain 1 case.

**Review by xvjiarui — APPROVED** [review]

File: tests/test_data/test_transform.py [hunk 16]
```diff
@@ -10,6 +10,26 @@
 from mmseg.datasets.builder import PIPELINES
 
 
+def test_align():
```
Comment (Junjun2016) [inline tests/test_data/test_transform.py:13]:
```suggestion
def test_resize_to_multiple():
```

File: configs/segformer/segformer_mit-b1_512x512_160k_ade20k.py [hunk 17]
```diff
@@ -0,0 +1,37 @@
+_base_ = [
+    '../_base_/models/segformer_mit-b0.py', '../_base_/datasets/ade20k.py',
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b1_512x512_160k_ade20k.py:2]: Inherit from `configs/segformer/segformer_mit-b0_512x512_160k_ade20k.py`.

File: configs/segformer/segformer_mit-b1_512x512_160k_ade20k.py [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    _delete_=True,
+    policy='poly',
+    warmup='linear',
+    warmup_iters=1500,
+    warmup_ratio=1e-6,
+    power=1.0,
+    min_lr=0.0,
+    by_epoch=False)
+
+data = dict(samples_per_gpu=2)
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b1_512x512_160k_ade20k.py:37]: Can be inherit from `configs/segformer/segformer_mit-b0_512x512_160k_ade20k.py`?

File: configs/segformer/segformer_mit-b1_512x512_160k_ade20k.py [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    '../_base_/models/segformer_mit-b0.py', '../_base_/datasets/ade20k.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_160k.py'
+]
+
+# model settings
+model = dict(
+    pretrained='pretrain/mit_b1.pth',
+    backbone=dict(
+        embed_dims=64, num_heads=[1, 2, 5, 8], num_layers=[2, 2, 2, 2]),
+    decode_head=dict(in_channels=[64, 128, 320, 512], num_classes=150))
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b1_512x512_160k_ade20k.py:11]:
```suggestion
    decode_head=dict(in_channels=[64, 128, 320, 512]))
```

File: configs/segformer/segformer_mit-b2_512x512_160k_ade20k.py [hunk 20]
```diff
@@ -0,0 +1,37 @@
+_base_ = [
+    '../_base_/models/segformer_mit-b0.py', '../_base_/datasets/ade20k.py',
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b2_512x512_160k_ade20k.py:2]: Inherit from `configs/segformer/segformer_mit-b0_512x512_160k_ade20k.py`.

Comment (Junjun2016) [reply configs/segformer/segformer_mit-b2_512x512_160k_ade20k.py:2]: The same as above.

File: configs/segformer/segformer_mit-b0_512x512_160k_ade20k.py [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    _delete_=True,
+    policy='poly',
+    warmup='linear',
+    warmup_iters=1500,
+    warmup_ratio=1e-6,
+    power=1.0,
+    min_lr=0.0,
+    by_epoch=False)
+
+data = dict(samples_per_gpu=2)
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b0_512x512_160k_ade20k.py:33]: workers_per_gpu=2

File: configs/segformer/segformer_mit-b3_512x512_160k_ade20k.py [hunk 22]
```diff
@@ -0,0 +1,37 @@
+_base_ = [
+    '../_base_/models/segformer_mit-b0.py', '../_base_/datasets/ade20k.py',
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b3_512x512_160k_ade20k.py:2]: The same as aforementioned.

File: configs/segformer/segformer_mit-b4_512x512_160k_ade20k.py [hunk 23]
```diff
@@ -0,0 +1,37 @@
+_base_ = [
+    '../_base_/models/segformer_mit-b0.py', '../_base_/datasets/ade20k.py',
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b4_512x512_160k_ade20k.py:2]: The same as aforementioned.

File: configs/segformer/segformer_mit-b5_512x512_160k_ade20k.py [hunk 24]
```diff
@@ -0,0 +1,38 @@
+_base_ = [
+    '../_base_/models/segformer_mit-b0.py', '../_base_/datasets/ade20k.py',
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b5_512x512_160k_ade20k.py:2]: The same as aforementioned.

File: configs/segformer/segformer_mit-b5_640x640_160k_ade20k.py [hunk 25]
```diff
@@ -0,0 +1,91 @@
+_base_ = [
+    '../_base_/models/segformer_mit-b0.py', '../_base_/default_runtime.py',
```
Comment (Junjun2016) [inline configs/segformer/segformer_mit-b5_640x640_160k_ade20k.py:2]: Inherit from `configs/segformer/segformer_mit-b5_512x512_160k_ade20k.py`

Comment (Junjun2016) [reply configs/segformer/segformer_mit-b5_640x640_160k_ade20k.py:2]: After inherited, only need overwrite `pipeline` in `data`.

File: mmseg/models/decode_heads/segformer_head.py [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+from mmseg.models.decode_heads.decode_head import BaseDecodeHead
+from mmseg.ops import resize
+
+
+@HEADS.register_module()
+class SegformerHead(BaseDecodeHead):
+    """The all mlp Head of segformer.
+
+    This head is the implementation of
+    `Segformer <https://arxiv.org/abs/2105.15203>`
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/segformer_head.py:15]: > This head is the implementation of `DNLNet
>     <https://arxiv.org/abs/2006.06668>`_.

Missing `_`, please refer to other heads.

File: tools/model_converters/mit_convert.py [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+            new_v = v
+        else:
+            new_k = k
+            new_v = v
+        new_ckpt[new_k] = new_v
+    return new_ckpt
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
```
Comment (Junjun2016) [inline tools/model_converters/mit_convert.py:57]:
```suggestion
    parser = argparse.ArgumentParser(description='')
```

It is better to add some descriptions or docstring about this script.

**Review by Junjun2016 — CHANGES_REQUESTED** [review]

**Review by Junjun2016 — APPROVED** [review]
